dftimewolf/lib/state.py

```diff
@@ -614,3 +600,157 @@ def CheckErrors(self, is_global: bool=False) -> None:
+      self, module: ThreadAwareModule
+    ) -> List[Future]: # type: ignore
+    """Runs Process of a single ThreadAwareModule module.
+
+    Args:
+      module: The module that will have Process(container) called in a threaded
+          fashion."""
+    cont_count = len(self.GetContainers(module.GetThreadOnContainerType()))
+    logger.info(
+       f'Running {cont_count} threads, max {module.GetThreadPoolSize()} '
+       f'simultaneous for module {module.name}')
+
+    self.cursesdm.SetThreadedModuleContainerCount(module.name, cont_count)
+    self.cursesdm.UpdateModuleStatus(module.name, cdm.Status.PROCESSING)
+
+    futures = []
+
+    with ThreadPoolExecutor(max_workers=module.GetThreadPoolSize()) \
+        as executor:
+      pop = not module.KeepThreadedContainersInState()
+      for c in self.GetContainers(module.GetThreadOnContainerType(), pop):
+        futures.append(
+            executor.submit(
+                self._WrapThreads, module.Process, c, module.name))
+
+    return futures
+
+  def _RunModulePreProcess(self, module: ThreadAwareModule) -> None:
+    """Runs PreProcess of a single module.
+
+    Args:
+      module: The module that will have PreProcess() called."""
+    self.cursesdm.UpdateModuleStatus(module.name, cdm.Status.PREPROCESSING)
+    module.PreProcess()
+    self.cursesdm.UpdateModuleStatus(module.name, cdm.Status.PENDING)
+
+  def _RunModulePostProcess(self, module: ThreadAwareModule) -> None:
+    """Runs PostProcess of a single module.
+
+    Args:
+      module: The module that will have PostProcess() called."""
+    self.cursesdm.UpdateModuleStatus(module.name, cdm.Status.POSTPROCESSING)
+    module.PostProcess()
+    self.cursesdm.UpdateModuleStatus(module.name, cdm.Status.COMPLETED)
+
+  def _HandleFuturesFromThreadedModule(
+      self,
+      futures: List[Future],  # type: ignore
+      runtime_name: str) -> None:
+    """Handles any futures raised by the async processing of a module.
+
+    Args:
+      futures: A list of futures, returned by RunModuleProcessThreaded().
+      runtime_name: runtime name of the module."""
+    for fut in futures:
+      if fut.exception():
+        self.cursesdm.SetError(runtime_name, str(fut.exception()))
+        raise fut.exception()  # type: ignore
+
+  def _WrapThreads(self,
+                  process: Callable[[AttributeContainer], None],
+                  container: AttributeContainer,
+                  module_name: str) -> None:
+    """Wraps a ThreadPoolExecutor call to module.process with the
+    CursesDisplayManager status update methods.
+
+    Args:
+      process: A callable method: Process, belonging to a ThreadAwareModule.
+      container: The Container being processed by the thread.
+      module_name: The runtime name of the module."""
+
+    thread_id = threading.current_thread().getName()
+    container_name = (container.name  # type: ignore[attr-defined]
+        if hasattr(container, 'name') else str(container))
```

Comment:
Why not just encode this logic `container.__str__`? I like the idea that each container defines how it's being displayed to a user.

Reply:
Makes sense.